elf_reader.go

```diff
@@ -554,8 +556,11 @@ func (ec *elfCode) loadDataSections(maps map[string]*MapSpec, dataSections map[e
 
 		switch sec.Name {
 		case ".rodata":
-			mapSpec.Flags = unix.BPF_F_RDONLY_PROG
-			mapSpec.Freeze = true
+			// Only freeze and mark read-only if the kernel supports it.
+			if haveMapMutabilityModifiers() == nil {
```

Comment:
Putting both `Flags` and `Freeze` assignments behind `haveMapMutabilityModifiers()` causes the following test to fail:

```
TestLoadCollectionSpec/loader-clang-9-el.elf: elf_reader_test.go:97: program xdp_prog: can't load program: invalid argument: unrecognized bpf_ld_imm64 insn
```

`bpf_ld_imm64` should be used to load constants only, and the verifier in 4.19 seems to trip over loads from sections that are not frozen. Since the instruction was [introduced in 3.18](https://github.com/torvalds/linux/commit/02ab695bb37ee9ad515df0d0790d5977505dd04a), and `BPF_F_RDONLY_PROG` was introduced along with `BPF_MAP_FREEZE` in 5.2, what would be the correct behaviour here to support loading this program?

Reply:
/cc @lmb @joestringer @borkmann @pchaigno @qmonnet

Reply:
Looks like the `LoadMapValue` generated by clang-9 are only supported since kernel 5.2 ([d8eca5bbb2be](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=d8eca5bbb2be) ("bpf: implement lookup-free direct value access for maps")). The test is skipped (so doesn't fail) on 4.9. Not sure how the instruction appeared by moving the assignments behind the guard, though? Do you know if these `LoadMapValue` instructions appeared before in the programs produced for the tests on 4.19 before the change?

Reply:
You can't use data sections on 4.19. Setting flags here means that the check in createMap returns an internal.UnsupportedFeatureError, which the test suite knows to skip.

Reply:
Found the root cause in thread https://github.com/cilium/ebpf/pull/94/files#r424351006, thanks!